### ISSUE TITLE
A suggestion for future direction

### ISSUE THREAD

**Issue by xuancong84** [dmlc/mxnet#4864]
If you are an experienced neural network programmer, you should know what is missing in the current state-of-the-art deep learning (DL) toolkits such as Theano, Caffein, Tensorflow, etc. So I am writing this post to give you guys some suggestion for the MXNET development. And I hope to see some light in MXNET which is a new and growing neural network (NN) toolkit.

Core Issues:
0. Coding efficiency.
Einstein has said, if two theories can both explain the same phenomenon, we typically adopt the simpler one. Analogously, if two NN programming languages can implement the same neural network, we prefer the one with a simpler code. This remind me one of the remarkable difference between Theano and Tensorflow, where you can use X[:,2:4,0:2,:] extract a slice from a rank-4 Tensor, but you have to call a complicated slicing function in Tensorflow. For MXNET, I realize that many of your API functions requires a very long prefix, e.g., 
```
net = mx.symbol.FullyConnected(data=net, name='fc1', num_hidden=128)
net = mx.symbol.Activation(data=net, name='tanh1', act_type="tanh")
```
I just want to ask you a simple question, how complex can it be just to write down `y=tanh(W*x+b)`? Do you really need to make things that complicated, just to achieve the purpose mentioned in the next point?

1. Auto-batching.
Whenever we want to implement a NN from a research paper, most of the effort we spend is in fact in doing the batching. The original mathematical equations are in fact quite straight forward to implement, it is the batching which is troublesome, you may refer to the paper titled "deep learning with dynamic computation graphs" which shows some on-going effort in the Google DL team. Up to today, no state-of-the-art NN toolkit can automatically handle batching. This might sound good for you.

2. Dynamic computation graph.
The ability to change computation graph dynamically during runtime is crucial for future artificial intelligence. This is because even for human and animals, the structure of the brain's neural network  changes as the body grow up and learns. We can see that nature does not evolve species with fixed-structure neural networks, this implies that all static-computation-graph NN toolkits must somehow have some fundamental limitations, which is still not very clear at state-of-the-art level of research. Up to today, only MXNET and DyNet support dynamic computation graph (also known as imperative programming, which "define by run"), most NN toolkits only support static computation graph (also known as declarative programming, which "define and run"). There are two issues here: 1. performance efficiency, we all know that declarative programming can compile the graph to make it run faster, however, nowadays we have the JIT (just-in-time) compiler concept, if NN toolkit can be designed to dynamically compile the computation graph while running the graph, then it will be great for imperative programming to catch up with declarative programming in terms of execution efficiency; 2. modelling capability, can we say that imperative programming is a functional superset of declarative programming? Or in another word, whatever tasks that can be achieved in declarative programming can also be achieved in imperative programming, but not vice versa? At least I believe so, a dynamically varying structure can achieve something that a fixed-structure network cannot achieve because given the data, you do not know what network structure is optimal for this task. You need to try all possible structures for a fixed-structure network which there are too many, but the dynamic network will learn the structure instead.
For MXNET, is it really necessary to achieve both declarative programming and imperative programming? If one is a superset of the other, why not just focus on the superset?

In conclusion, I don't believe MXNET is a very promising DL toolkit given its current API structure, unless there is some big leap breaking changes.
